File: scala-code-mike/src/main/scala/funar/Cards.scala
package funar

sealed trait Suit
object Suit {
  case object Diamonds extends Suit
  case object Clubs extends Suit
  case object Spades extends Suit
  case object Hearts extends Suit

  val all = Seq(Diamonds, Clubs, Spades, Hearts)
}

sealed trait Rank {
  val value: Int
}

object Rank {
  case object Two extends Rank { override val value = 2 }
  case object Three extends Rank { override val value = 3 }
  case object Four extends Rank { override val value = 4 }
  case object Five extends Rank { override val value = 5 }
  case object Six extends Rank { override val value = 6 }
  case object Seven extends Rank { override val value = 7 }
  case object Eight extends Rank { override val value = 8 }
  case object Nine extends Rank { override val value = 9 }
  case object Ten extends Rank { override val value = 10 }
  case object Jack extends Rank { override val value = 11 }
  case object Queen extends Rank { override val value = 12 }
  case object King extends Rank { override val value = 13 }
  case object Ace extends Rank { override val value = 14 }

  val all = Seq(Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten,
                           Jack, Queen, King, Ace)
}

File: mike/hearts/src/Cards.hs
-- Das französische Blatt
module Cards where

import qualified Data.Set as Set
import Data.Set (Set)

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

-- "Make illegal states unrepresentable." - Yaron Minsky

-- Illegale Werte gar nicht herstellen.

-- 1. Typsystem
-- 2. Validierung

data Suit = Diamonds | Clubs | Spades | Hearts
    deriving (Show, Eq)

allSuits :: [Suit]
allSuits = [Diamonds, Clubs, Spades, Hearts]

data Rank =
    Two | Three | Four | Five | Six | Seven | Eight
  | Nine | Ten | Jack | Queen | King | Ace
  deriving (Show, Eq, Ord)

allRanks :: [Rank]
allRanks =
  [ Two
    , Three
    , Four
    , Five
    , Six
    , Seven
    , Eight
    , Nine
    , Ten
    , Jack
    , Queen
    , King
    , Ace
  ]

data Card = Card { suit :: Suit, rank :: Rank }
  deriving Show

cartesianProduct :: [a] -> [b] -> [(a, b)]
cartesianProduct as bs =
    concat (map (\ a -> map (\b -> (a, b)) bs) as)

deck :: [Card]
deck = map (uncurry Card) (cartesianProduct allSuits allRanks)

-- >>> length deck
-- 52

cardOrder :: Card -> Card -> Maybe Ordering
cardOrder card1 card2 =
    if suit card1 == suit card2
    then Just (compare (rank card1) (rank card2))
    else Nothing
